Add ability to exclude bookmarked element

commelius-grela opened the pull request:
Allows to exclude the bookmarked element by checking it's `_bookmarking` property and skipping if disabled. Last accepted (or reset) id will be used.

commelius-grela committed:
Add ability to exclude bookmarked element

Allows to exclude the bookmarked element by checking it's `_bookmarking` property and skipping if disabled. Last accepted (ore reset) id will be used.

commelius-grela committed:
Add ability to exclude bookmarked element

commelius-grela committed:
Merge pull request #1 from commelius/commelius-exclude

Add ability to exclude bookmarked element

moloko commented:
Nice idea... it will need an update to properties.schema as well though to allow the user to set the `_bookmarking` property of articles/blocks/components

commelius-grela committed:
Add ability to exclude bookmarked element

Modified the schema for article/block/component to add the `_bookmarking` property

LukaszGrela commented:
Like that?

brian-learningpool commented:
+1, not tested it yet, but the code looks fine.

oliverfoster commented:
+1

moloko reviewed (changes requested):
Well, first of all huge apologies for abandoning this for such a long time!
I've just given this a test in the framework and it works really well.
I just have a problem with the properties.schema - the default setting is false which I think means that `_bookmarking._isEnabled` will be set to `false` on every single article, block and component automatically!

I wonder if a better way to do this would be to have an `_exclude` property on the `_bookmarking` settings in contentObjects.json? So, assuming you were bookmarking at block level you could set `"_exclude": "b-05, b-10, b-15"`?

LukaszGrela commented:
Let me see:)

commelius-grela commented:
The idea was to indicate per object (article/block/component) and easily exclude it by inserting the 
```
_bookmarking: {
_isEnabled:false
}
```
if it is `true` or didn't exist then it should be allowing to bookmark (default), [this line](https://github.com/adaptlearning/adapt-contrib-bookmarking/pull/22/files#diff-dde751301dd9e087f3f642f2573c9d43R164) should be positive only when entry exists, has the `_isEnabled` and it is set to `false`.

You are proposing to have a single settings location in the `contentObjects.json` with `_exclude` property and put the ids for exlusion? It may work too, this approach was just quicker to implement:)

moloko commented:
Sorry, I didn't explain this very well - what I mean is that `default: false` property on lines [123](https://github.com/adaptlearning/adapt-contrib-bookmarking/pull/22/files#diff-4517165ffe55c65dfed48457eadd2d1cR123),  [144](https://github.com/adaptlearning/adapt-contrib-bookmarking/pull/22/files#diff-4517165ffe55c65dfed48457eadd2d1cR144) and [165](https://github.com/adaptlearning/adapt-contrib-bookmarking/pull/22/files#diff-4517165ffe55c65dfed48457eadd2d1cR165) of properties.schema will cause the authoring tool to add
```
_bookmarking: {
   _isEnabled:false
}
```
to _every_ article, block and component in the course, effectively disabling bookmarking completely unless you go through every article/block/component changing it back to `true`... changing these lines to `default: true` will make it work as you intend.
> You are proposing to have a single settings location in the contentObjects.json with _exclude property and put the ids for exclusion? 

Yes, exactly. I was just considering the way that the AT will add a `_bookmarking._isEnabled` property to _every_ article, block & component within the course - this seemed like quite a lot to add when, in general, disabling bookmarking at this sort of level is something that's only occasionally required... so I thought that having just one property at the content object level would be a more efficient way of doing it.

What I would say is that using `"inputType": "Checkbox"` for boolean properties like `_isEnabled` looks a lot nicer in the AT and is more usable - just one click to change the setting instead of two.

moloko commented:
closing due to inactivity
